src/include/duckdb/execution/operator/join/physical_hash_join.hpp

```diff
@@ -9,22 +9,60 @@
 #pragma once
 
 #include "duckdb/common/types/chunk_collection.hpp"
+#include "duckdb/common/value_operations/value_operations.hpp"
 #include "duckdb/execution/join_hashtable.hpp"
+#include "duckdb/execution/operator/join/perfect_hash_join_executor.hpp"
 #include "duckdb/execution/operator/join/physical_comparison_join.hpp"
 #include "duckdb/execution/physical_operator.hpp"
 #include "duckdb/planner/operator/logical_join.hpp"
 
 namespace duckdb {
 
+class PhysicalHashJoinState : public PhysicalOperatorState {
```

Comment:
We generally prefer to keep these in the `.cpp` files to keep implementation details out of headers. The structures also seem to contain both information required by the regular hash join and information required by the perfect hash join.

Could we move these back to the `.cpp` file? We could then create states for the perfect hash join (`PerfectHashJoinState`, etc) that are stored as child states of these states. Those states could live entirely in `perfect_hash_join_executor.cpp` and be created using e.g. `PerfectHashJoinExecutor::GetState`.

Reply:
Ok. I didn't know there was a reason for that.